File: easy/0783_minimum_distance_between_bst_nodes/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <stack>
#include <vector>

class Solution {
public:
  int minDiffInBST(utils::TreeNode *root) {
    std::stack<utils::TreeNode *> s;
    std::vector<int> values;

    utils::TreeNode *curr = root;

    while (!s.empty() || curr) {
      while (curr) {
        s.push(curr);
        curr = curr->left;
      }

      curr = s.top();
      s.pop();

      values.push_back(curr->val);
      curr = curr->right;
    }

    int minDiff = INT_MAX;
    for (size_t i = 0; i < values.size() - 1; ++i)
      minDiff = std::min(minDiff, abs(values[i] - values[i + 1]));

    return minDiff;
  }
};

File: easy/0783_minimum_distance_between_bst_nodes/solution.py
from utils.python.nodes import TreeNode


class Solution:
    def minDiffInBST(self, root: TreeNode | None) -> int:
        stack = []
        values = []

        curr = root

        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left

            curr = stack.pop()
            values.append(curr.val)
            curr = curr.right

        return min(
            [abs(values[i] - values[i + 1]) for i in range(len(values) - 1)]
        )
